tests/monitoring_test.go

```diff
@@ -0,0 +1,534 @@
+package tests_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	promv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+
+	"kubevirt.io/containerized-data-importer/pkg/controller"
+	"kubevirt.io/containerized-data-importer/pkg/storagecapabilities"
+	"kubevirt.io/containerized-data-importer/tests/framework"
+	"kubevirt.io/containerized-data-importer/tests/utils"
+)
+
+const (
+	dataVolumeName      = "test-dv"
+	nonExistentImageURL = "http://www.example.com/example.img"
+)
+
+var _ = Describe("[Destructive] Monitoring Tests", func() {
+	f := framework.NewFramework("monitoring-test")
+
+	var (
+		cr                     *cdiv1.CDI
+		crModified             bool
+		cdiPods                *corev1.PodList
+		numAddedStorageClasses int
+		originalMetricVal      int
+	)
+
+	BeforeEach(func() {
+		if !f.IsPrometheusAvailable() {
+			Skip("This test depends on prometheus infra being available")
+		}
+
+		cr = getCDI(f)
+		cdiPods = getCDIPods(f)
+		originalMetricVal = getMetricValue(f, "kubevirt_cdi_incomplete_storageprofiles_total")
+	})
+
+	AfterEach(func() {
+		By("Delete unknown storage classes")
+		for i := 0; i < numAddedStorageClasses; i++ {
+			name := fmt.Sprintf("unknown-sc-%d", i)
+			_, err := f.K8sClient.StorageV1().StorageClasses().Get(context.TODO(), name, metav1.GetOptions{})
+			if err != nil && errors.IsNotFound(err) {
+				continue
+			}
+			err = f.K8sClient.StorageV1().StorageClasses().Delete(context.TODO(), name, metav1.DeleteOptions{})
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		if crModified {
+			removeCDI(f, cr)
+			ensureCDI(f, cr, cdiPods)
+			crModified = false
+		}
+
+		Eventually(func() int {
+			return getMetricValue(f, "kubevirt_cdi_incomplete_storageprofiles_total")
+		}, 5*time.Minute, 5*time.Second).Should(BeNumerically("==", originalMetricVal))
+	})
+
+	Context("[rfe_id:7101][crit:medium][vendor:cnv-qe@redhat.com][level:component] Metrics tests", func() {
+		var (
+			metricPollingInterval = 5 * time.Second
+			metricPollingTimeout  = 5 * time.Minute
```

Comment:
Can move these a level up and use them for https://github.com/kubevirt/containerized-data-importer/blob/ca32eba41c93728bb16b607c169054e8c2f60855/tests/monitoring_test.go#L79